staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go

```diff
@@ -0,0 +1,261 @@
+	}
+
+	if existing, groupExists := rdm.apiGroups[groupName]; groupExists {
+		// If this version already exists, replace it
+		versionExists := false
+
+		// Not very efficient, but in practice there are generally not many versions
+		for i := range existing.Versions {
+			if existing.Versions[i].Version == value.Version {
+				existing.Versions[i] = value
+				versionExists = true
+				break
+			}
+		}
+
+		if !versionExists {
+			existing.Versions = append(existing.Versions, value)
+		}
+
+	} else {
+		group := &apidiscoveryv2beta1.APIGroupDiscovery{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "APIGroupDiscovery",
+				APIVersion: "apidiscovery.k8s.io/v2beta1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: groupName,
+			},
+			Versions: []apidiscoveryv2beta1.APIVersionDiscovery{value},
+		}
+		rdm.apiGroups[groupName] = group
+		rdm.apiGroupNames = append(rdm.apiGroupNames, group.Name)
+
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cache.Store(nil)
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroupVersion(apiGroup metav1.GroupVersion) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	group, exists := rdm.apiGroups[apiGroup.Group]
+	if !exists {
+		return
+	}
+
+	for i := range group.Versions {
+		if group.Versions[i].Version == apiGroup.Version {
+			group.Versions = append(group.Versions[:i], group.Versions[i+1:]...)
+			break
+		}
+	}
+
+	if len(group.Versions) == 0 {
+		for i, g := range rdm.apiGroupNames {
+			if g == group.Name {
+				rdm.apiGroupNames = append(rdm.apiGroupNames[:i], rdm.apiGroupNames[i+1:]...)
+			}
+		}
+		delete(rdm.apiGroups, group.Name)
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cache.Store(nil)
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroup(groupName string) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	delete(rdm.apiGroups, groupName)
+
+	// Reset response document so it is recreated lazily
+	rdm.cache.Store(nil)
+}
+
+// Prepares the api group list for serving by converting them from map into
+// list and sorting them according to insertion order
+func (rdm *resourceDiscoveryManager) calculateAPIGroups() []apidiscoveryv2beta1.APIGroupDiscovery {
+	rdm.lock.RLock()
+	defer rdm.lock.RUnlock()
+
+	// Re-order the apiGroups by their insertion order
+	groups := []apidiscoveryv2beta1.APIGroupDiscovery{}
+	for _, v := range rdm.apiGroupNames {
+		groups = append(groups, *rdm.apiGroups[v].DeepCopy())
+	}
+	return groups
+}
+
+func (rdm *resourceDiscoveryManager) generateCache() *cachedGroupList {
+	response := apidiscoveryv2beta1.APIGroupDiscoveryList{
+		Items: rdm.calculateAPIGroups(),
+	}
+	etag, err := CalculateETag(response)
+	if err != nil {
+		klog.Errorf("failed to calculate etag for discovery document: %s", etag)
+	}
+
+	cached := &cachedGroupList{
+		cachedResponse:     response,
+		cachedResponseETag: etag,
+	}
+
+	cacheLoad := rdm.cache.Load()
+	// Another thread may have loaded the cache while calculating the
```

Comment:
without a lock, could we end up with

1. thread 1 adds GV=foo/v1 so we set to nil
2. thread 2 hits serveHTTP and observes a cache nil and calls generateCache, it hangs just after the conditional and before calling the store
3. thread 3 adds another GV=bar/v2, so we set to nil
4. thread 4 hits serveHTTP and observes a cache nil and calls generateCache, it sees nil and sets foo/v1,bar/v2
5. thread 2 gets released and stores foo/v1 (notice bar is missing).

Reply:
Yep, sounds like the sort of problems we've solved with etcd by having a resourceVersion. We should solve it the exact same way by doing [a compare and swap](https://pkg.go.dev/sync/atomic#Pointer.CompareAndSwap). The Golang documentation for that function is horrible if you're not sure what you're doing, so you can look at https://en.wikipedia.org/wiki/Compare-and-swap.

Reply:
I think that still leaves a bug:
1. thread 1 adds GV=foo/v1 so we set to nil
2. thread 2 hits serveHTTP and observes a cache nil and calls generateCache, it hangs just after the conditional and before calling the store
3. thread 3 adds another GV=bar/v2, so we set to nil
5. thread 2 gets released and stores foo/v1 (notice bar is missing).

Basically, we don't detect the "transition" from nil to nil, done in 3. Would it work if we do the compare and swap while holding the read lock?

Reply:
Thanks for the pointer Antoine. Yeah we can't detect transitions from nil to nil. I'm going with a workaround to initialize an empty `&cacheGroupList{}` for resetting so the transition between nil to nil is captured.

Reply:
> Thanks for the pointer Antoine. Yeah we can't detect transitions from nil to nil. I'm going with a workaround to initialize an empty &cacheGroupList{} for resetting so the transition between nil to nil is captured.

reminder to @apelisse to double check.  I think the new impl closes the gap I pointed out.